### DISCUSSION TITLE
Window freez on type text into TFileEditor

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#114] [Q&A] [answered]
Hello,
I oversee something, the Code compiled fine, but on runtime, the Application can not be close (Alt+X), and the Window can not be closed. Please Help.
I have the following Code:
```
void TVDemo::tvEditor()
{
    TMyTvEditor *editor = (TMyTvEditor *) validView(new TMyTvEditor(TRect(0, 0, 79, 23)));

    if (editor != 0) {
        editor->helpCtx = hcCalendar;
        deskTop->insert( editor );
    }
}

class TMyTvEditor : public TWindow {
public:
    TMyTvEditor(const TRect& bounds):
        TWindowInit(&TMyTvEditor::initFrame),
        TWindow(bounds, "Mein Fenster", wnNoNumber) {
            
        // Viewer erstellen
        viewer = new TScroller(TRect(2, 2, 39, 23),
            standardScrollBar(sbHorizontal),
            standardScrollBar(sbVertical));
        insert(viewer);

        // Text-Editor erstellen
        TScrollBar *hScrollBar = new TScrollBar( TRect( 2, size.y - 1, size.x - 24, size.y ) );
        hScrollBar->hide();
        insert(hScrollBar);

        TScrollBar *vScrollBar = new TScrollBar( TRect( size.x - 23, 1, size.x - 22, size.y - 1 ) );
        vScrollBar->hide();
        insert(vScrollBar);

        editor = new TFileEditor(TRect(1, 1, size.x - 24, size.y - 1), hScrollBar, vScrollBar, nullptr, "www.txt");
        insert(editor);

        options |= ofSelectable;
    }

    void handleEvent(TEvent &event) {
        TWindow::handleEvent(event);
        if (event.what == evCommand) {
            switch (event.message.command) {
                case cmClose:
                    close();
                    clearEvent(event);
                    break;
                case cmButton1:
                    // Aktion für Schaltfläche 1
                    break;
                case cmButton2:
                    // Aktion für Schaltfläche 2
                    break;
            }
        }
        else {
            clearEvent(event);
        }
    }

private:
    TScroller *viewer;
    TFileEditor *editor;
    TButton *button1;
    TButton *button2;
};

```

**Comment by magiblot** [comment, marked as answer]
Hi Jens!

Given that this issue is related to the usage of the library and it is not a bug, I have converted it into a GitHub discussion.

There are two problem with your code.

In the first place, the `handleEvent` method of `TMyTvEditor` is currently calling `clearEvent(event)` on events that this view isn't handling. A view must only clear events that it was intended to handle, as it will otherwise prevent other views from receiving these events.

Therefore, that part of the code must be edited as follows:
```diff
                case cmButton2:
                    // Aktion für Schaltfläche 2
                    break;
            }
        }
-       else {
-           clearEvent(event);
-       }
    }
```

In the second place, the reason why exiting the program doesn't work is because `TVDemo`, which is a `TGroup`, first validates that it is OK to exit:

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/source/tvision/tgroup.cpp#L185

This validation reaches `TFileEditor`, which checks that there are no unsaved changes in the editor buffer:

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/source/tvision/tfiledtr.cpp#L270

`TFileEditor` then calls `TEditor::editorDialog`, which is meant to ask the user what to do in this case:

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/source/tvision/tfiledtr.cpp#L278

However, your application does not define a custom `TEditor::editorDialog`, so the default one is used:

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/source/tvision/editstat.cpp#L18-L21

As a result, nothing is shown and it is automatically assumed that the file cannot be saved. Therefore, the program does not exit.

To solve this, you should define your own `TEditor::editorDialog`, like the TVEdit application does:

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/examples/tvedit/tvedit3.cpp#L106

https://github.com/magiblot/tvision/blob/e9fac379992815aaadcba99f55d8d34b7037a311/examples/tvedit/tvedit1.cpp#L62

Cheers.

**Reply by paule32** [reply]
This solved my issue. Thanks !
